### DISCUSSION TITLE
Application level component with SignalR connection

### DISCUSSION THREAD

**Discussion by vzdesic** [oqtane/oqtane.framework#1441] [Q&A] [answered]
I would like to ask you guys about opinions related to my challenge and possible solutions inside the Oqtane framework.

I want to have a small area of the application dedicated to receive notifications from my backend. My plan is to have Runtime parameter set to WebAssembly in my application. Notifications can come from multiple services. Mostly, notifications will just be displayed to the user, but there will also be such notifications which will trigger modal form or popup filled with data. Notifications will be sent over SignalR channel from the backend, or at least this is the plan for now. Another option is to have a singleton object in the Oqtane subscribed to some queue or topic where notifications will be published, but my questions are not related to that option because I think I understand how to implement that. I will have to register a singleton from the external module by using IClientStartup and subscribe to the queue inside the constructor, so this is clear, I hope.

My questions are following:
- Where to put this common component / control, shown on every page, and dedicated to show the application level notifications? From my point of view, Oqtane Theme should contain such common components shown on every page of the application. I hope I don't have to change any Oqtane framework source code to accomplish that, right? My confusion comes from the fact that Oqtane Theme is loaded on every request, and I am trying to see if this can be an issue to keep the SignalR connection between different requests.
- How to properly handle modal form or popup in the Oqtane which can be shown anywhere in the application? It seems that external modules cannot serve this purpose, because it will have to be imported on every page in this case.
- Do you see any issue related to keeping SignalR connection across the application?

Thanks a lot,
Vedran

**Comment by sbwalker** [comment, marked as answer]
I have not had time to experiment with SignalR on Blazor WebAssembly yet. I assume you have already seen the documentation from Microsoft on this topic which seems to indicate it is fairly straightforward:

https://docs.microsoft.com/en-us/aspnet/core/tutorials/signalr-blazor?view=aspnetcore-5.0&tabs=visual-studio&pivots=webassembly

From an Oqtane perspective there are a few options:

- create a Theme Control and include your Theme Control in your Theme so that it is present on all pages.
- create a Module and then use the option Show On All Pages to have the module injected into a Theme Pane at runtime

In both cases you need to keep in mind that when running on WebAssembly there is only a single page request to the server when the browser first connects, and after that the only interactions with the server are via WebAPI requests retrieving data - there are no page requests. So the SignalR connection should be preserved.
